ul/src/association/client.rs

```diff
@@ -0,0 +1,396 @@
+                transfer_syntaxes: transfer_syntax_uids
+                    .iter()
+                    .map(|uid| uid.to_string())
+                    .collect(),
+            })
+            .collect();
+        let msg = Pdu::AssociationRQ {
+            protocol_version,
+            calling_ae_title: calling_ae_title.to_string(),
+            called_ae_title: called_ae_title.to_string(),
+            application_context_name: application_context_name.to_string(),
+            presentation_contexts: presentation_contexts.clone(),
+            user_variables: vec![],
+        };
+
+        let mut socket = std::net::TcpStream::connect(address).context(Connect)?;
+
+        // send request
+        write_pdu(&mut socket, &msg).context(SendRequest)?;
+
+        // receive response
+        let msg = read_pdu(&mut socket, max_pdu_length).context(ReceiveResponse)?;
+
+        match msg {
+            Pdu::AssociationAC {
+                protocol_version: protocol_version_scp,
+                application_context_name: _,
+                presentation_contexts: presentation_contexts_scp,
+                user_variables: _,
+            } => {
+                ensure!(
+                    protocol_version == protocol_version_scp,
+                    ProtocolVersionMismatch {
+                        expected: protocol_version,
+                        got: protocol_version_scp,
+                    }
+                );
+
+                let selected_context = presentation_contexts_scp
+                    .into_iter()
+                    .find(|c| c.reason == PresentationContextResultReason::Acceptance)
+                    .context(NoAcceptedPresentationContexts)?;
+
+                let presentation_context = presentation_contexts
+                    .into_iter()
+                    .find(|c| c.id == selected_context.id)
+                    .context(NoAcceptedPresentationContexts)?;
+
+                Ok(ClientAssociation {
+                    presentation_context_id: selected_context.id,
+                    abstract_syntax_uid: presentation_context.abstract_syntax,
+                    transfer_syntax_uid: selected_context.transfer_syntax,
+                    max_pdu_length,
+                    socket,
+                })
+            }
+            Pdu::AssociationRJ { result, source } => Rejected {
+                association_result: result,
+                association_source: source,
+            }
+            .fail(),
+            pdu @ Pdu::AbortRQ { .. }
+            | pdu @ Pdu::ReleaseRQ { .. }
+            | pdu @ Pdu::AssociationRQ { .. }
+            | pdu @ Pdu::PData { .. }
+            | pdu @ Pdu::ReleaseRP { .. } => UnexpectedResponse { pdu }.fail(),
+            pdu @ Pdu::Unknown { .. } => UnknownResponse { pdu }.fail(),
+        }
+    }
+}
+
+/// A DICOM upper level association from the perspective
+/// of an association requester.
+#[derive(Debug)]
+pub struct ClientAssociation {
```

Comment:
Note to self: an SCU may want to use multiple presentation contexts during the same association. The API should change to support this.